sentry-android-core/src/main/java/io/sentry/android/core/SpanFrameMetricsCollector.java

```diff
@@ -133,34 +239,161 @@ public void onSpanFinished(final @NotNull ISpan span) {
-      final boolean isSlow,
-      final boolean isFrozen,
-      final float refreshRate) {
-
-    if (isFrozen) {
-      currentFrameMetrics.addFrozenFrame(durationNanos, delayNanos);
-    } else if (isSlow) {
-      currentFrameMetrics.addSlowFrame(durationNanos, delayNanos);
-    } else {
-      currentFrameMetrics.addNormalFrame(durationNanos);
-    }
-
-    lastRefreshRate = refreshRate;
+      long frameStartNanos,
+      long frameEndNanos,
+      long durationNanos,
+      long delayNanos,
+      boolean isSlow,
+      boolean isFrozen,
+      float refreshRate) {
+
+    // buffer is full, skip adding new frames for now
+    // once a span finishes, the buffer will trimmed
+    if (frames.size() > MAX_FRAMES_COUNT) {
+      return;
+    }
+
+    final long expectedFrameDurationNanos =
+        (long) ((double) ONE_SECOND_NANOS / (double) refreshRate);
+    lastKnownFrameDurationNanos = expectedFrameDurationNanos;
+
+    frames.add(
+        new Frame(
+            frameStartNanos,
+            frameEndNanos,
+            durationNanos,
+            delayNanos,
+            isSlow,
+            isFrozen,
+            expectedFrameDurationNanos));
+  }
+
+  private static int interpolateFrameCount(
+      final @NotNull SentryFrameMetrics frameMetrics,
+      final long frameDurationNanos,
+      final long spanDurationNanos) {
+    // if there are no content changes on Android, also no new frame metrics are provided by the
+    // system
+    // in order to match the span duration with the total frame count,
+    // we simply interpolate the total number of frames based on the span duration
+    // this way the data is more sound and we also match the output of the cocoa SDK
+    final long frameMetricsDurationNanos = frameMetrics.getTotalDurationNanos();
+    final long nonRenderedDuration = spanDurationNanos - frameMetricsDurationNanos;
+    if (nonRenderedDuration > 0) {
+      return (int) (nonRenderedDuration / frameDurationNanos);
+    }
+    return 0;
+  }
+
+  private static int addPendingFrameDelay(
+      @NotNull final SentryFrameMetrics frameMetrics,
+      final long frameDurationNanos,
+      final long spanEndNanos,
+      final long nextScheduledFrameNanos) {
+    final long pendingDurationNanos = Math.max(0, spanEndNanos - nextScheduledFrameNanos);
+    final boolean isSlow =
+        SentryFrameMetricsCollector.isSlow(pendingDurationNanos, frameDurationNanos);
+    if (isSlow) {
+      // add a single slow/frozen frame
+      final boolean isFrozen = SentryFrameMetricsCollector.isFrozen(pendingDurationNanos);
+      final long pendingDelayNanos = Math.max(0, pendingDurationNanos - frameDurationNanos);
+      frameMetrics.addFrame(pendingDurationNanos, pendingDelayNanos, true, isFrozen);
+      return 1;
+    }
+    return 0;
+  }
+
+  @ApiStatus.Internal
+  public interface FrameTimeProvider {
+    long now();
+  }
+
+  @ApiStatus.Internal
+  public static class NanoTimeStamp implements Comparable<NanoTimeStamp> {
```

Comment:
`l`: this also potentially could be replaced by the existing `SentryNanotimeDate`

Reply:
I agree, but we'd have to add a method to the `SentryDate` to get the current nano time, as the current implementation returns the date converted to nanoseconds (btw, is that intentional?)

Reply:
Actually, if we use dates only for differences (and we are here), we can use `SentryNanotimeDate` without problems

Reply:
I don't know why but `SentryNanotimeDate.nanoTimestamp()` only gives you millisecond precision, `diff` however works in nanos. Created a ticket here to follow up on this in the next sync: https://github.com/getsentry/sentry-java/issues/3210